File: codes/Hashtable.py
import matplotlib
import numpy as np
import matplotlib.pyplot as plt
from random import choice

class HashTable:
    def __init__(self, h, n=10):
        self.h = h
        self.n = n
        self.bucket = [[] for i in range(n)]

    def insert(self, x):
        self.bucket[self.h(x)].append(x)

    def delete(self, x):
        poss = self.bucket[self.h(x)]
        for i in range(len(poss)):
            if poss[i] == x:
                return poss.pop(i)
        return None

    def find(self, x):
        poss = self.bucket[self.h(x)]
        for i in range(len(poss)):
            if poss[i] == x:
                return poss[i]
        return None

def mi_Mod(x, n=10):
    return x % n

HT = HashTable(mi_Mod, 10)

x = 1234567
y = 76554334234
HT.insert(x)

print(HT.find(x))

def randomFn(x, n=10):
    return choice(range(n))

print(randomFn(52))
print(randomFn(3235235))

HT = HashTable(randomFn, 10)

x = 1234567
y = 76554334234
HT.insert(x)
print(HT.find(x))

print(randomFn(52))
print(randomFn(3235235))
print(randomFn(52))
print(randomFn(3235235))

def RandomHashFun(M, n=10):
    fnTable = [None for i in range(M)]
    for x in range(M):
        fnTable[x] = choice(range(n))

    def randomFn(x):
        return fnTable[x]

    return randomFn

randomFn2 = RandomHashFun(1000, 10)
print(randomFn2(52))
print(randomFn2(324))
print(randomFn2(52))
print(randomFn2(324))

HT = HashTable(randomFn2, 10)
x = 123
y = 76
HT.insert(x)

File: codes/Hashtable.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#define N 10

typedef struct {
    int (*h)(int, int);
    int n;
    int *bucket[N];
} HashTable;

int mi_Mod(int x, int n) {
    return x % n;
}

int insert(HashTable *ht, int x) {
    int index = ht->h(x, ht->n);
    if (ht->bucket[index] == NULL) {
        ht->bucket[index] = malloc(sizeof(int));
        *(ht->bucket[index]) = x;
    } else {
        int *temp = realloc(ht->bucket[index], sizeof(int));
        if (temp == NULL) {
            return 0; 
        } else {
            ht->bucket[index] = temp;
            *(ht->bucket[index]) = x;
        }
    }
    return 1;
}

int find(HashTable *ht, int x) {
    int index = ht->h(x, ht->n);
    if (ht->bucket[index] != NULL && *(ht->bucket[index]) == x) {
        return *(ht->bucket[index]);
    }
    return -1; 
}

int randomFn(int x, int n) {
    srand((unsigned int)time(NULL) + x);  
    return rand() % n;
}

int main() {
    HashTable HT;
    HT.h = &mi_Mod;
    HT.n = N;

    for (int i = 0; i < N; i++) {
        HT.bucket[i] = NULL;
    }

    int x = 1234567;
    long long int y = 76554334234;

    insert(&HT, x);

    printf("%d\n", find(&HT, x));

    srand((unsigned int)time(NULL)); 
    printf("%d\n", randomFn(52, N));
    printf("%d\n", randomFn(3235235, N));

    for (int i = 0; i < N; i++) {
        free(HT.bucket[i]);
    }

    return 0;
}
